u8timeseries/models/torch_forecasting_model.py

```diff
@@ -0,0 +1,505 @@
+        A PyTorch Dataset from a univariate TimeSeries.
+        The Dataset iterates a moving window over the time series. The resulting slices contain `(data, target)`,
+        where `data` and `target` are both 1-D sub-sequences of length [data_length]. The sequence contained in
+        target is shifted forward by [shift] positions, meaning that `target` contains the last
+        [data_length] - [shift] entries of `data` and then the [shift] following ones.
+
+        Parameters
+        ----------
+        series
+            The time series to be included in the dataset.
+        data_length
+            The length of the training and target sub-sequences.
+        shift
+            The number of positions that the target sequence is shifted forward compared to the training sequence.
+        """
+
+        self.series_values = series.values()
+        self.len_series = len(series)
+        self.data_length = len(series) - 1 if data_length is None else data_length
+        self.shift = shift
+
+        raise_if_not(self.data_length > 0,
+                     "The input sequence length must be positive. It is {}".format(self.data_length),
+                     logger)
+
+        raise_if_not(self.shift > 0,
+                     "The shift value must be positive. It is {}".format(self.data_length),
+                     logger)
+
+    def __len__(self):
+        return self.len_series - self.data_length - self.shift + 1
+
+    def __getitem__(self, index):
+        idx = index % self.__len__()
+        data = self.series_values[idx:idx + self.data_length]
+        target = self.series_values[idx + self.shift:idx + self.data_length + self.shift]
+        return torch.from_numpy(data).float().unsqueeze(1), torch.from_numpy(target).float().unsqueeze(1)
+
+
+class TorchForecastingModel(ForecastingModel):
+    # TODO: add init seed
+    # TODO: add is_stochastic & reset methods
+    # TODO: transparent support for multivariate time series
+    def __init__(self,
+                 batch_size: int = 32,
+                 output_length: int = 1,
+                 input_length: int = 10,
+                 n_epochs: int = 800,
+                 optimizer_cls: torch.optim.Optimizer = torch.optim.Adam,
+                 optimizer_kwargs: Dict = None,
+                 lr_scheduler_cls: torch.optim.lr_scheduler._LRScheduler = None,
+                 lr_scheduler_kwargs: Optional[Dict] = None,
+                 loss_fn: nn.modules.loss._Loss = nn.MSELoss(),
+                 model_name: str = "torch_model_run",  # TODO uid
+                 work_dir: str = os.getcwd(),
+                 log_tensorboard: bool = False,
+                 nr_epochs_val_period: int = 10,
+                 torch_device_str: Optional[str] = None):
+        """ Pytorch-based Forecasting Model.
+
+        This class is meant to be inherited to create a new pytorch-based forecasting module.
+        When subclassing this class, please make sure to set the self.model attribute
```

Comment:
The reason for why I opted to create the pytorch module in the __init__ function instead of an abstract method is because the model-specific parameters would all have to be passed to that method. So for instance, for the TCN the variable `dilation`would be an argument to the init function, which would then have to be forwarded in some way to the invocation of `model`, which would probably occur somewhere in the TorchForecastingModel superclass. I feel like this would get messy, no?

Reply:
I think you could maybe keep the creation of the pytorch module in init (as of now), but just implement an abstract method in all the sub-classes:
```
get_model():
    return self.model
```
By making it abstract you force the children classes to implement it (and you don't rely on the next person reading the comment to do it right). The when using the model you can call `self.get_model()` instead of `self.model`.
Does it make sense?

Reply:
I see how this would be cleaner in terms of subclass requirements, but the problem with this that I see is that the model hyperparameters passed to the `__init__`function of TorchForecastingModel will be required by the `get_model()` function. And since different subclasses can have different hyperparameters, wouldn't the `get_model()`function signature be inconsistent?

Reply:
OK, fair enough, let's leave it like that then.